File: intersection_of_2_lists.cpp
// Time Complexity : O(m+n) 
// m-> length of A, n->length of B
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
// Calculate the length of list A and list B. Adjust the longer list to have same #of nodes 
// as the smaller list. Then check the intersection point of both the lists.

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int lenA = 0;
        int lenB = 0;

        ListNode* nodeA = headA;
        ListNode* nodeB = headB;

        // Get length of A
        while(nodeA != nullptr)
        {
            lenA++;
            nodeA= nodeA->next;
        }
        
        // Get length of B
        while(nodeB != nullptr)
        {
            lenB++;
            nodeB= nodeB->next;
        }

        // Update the head to match the length of A and B
        while(lenA > lenB)
        {
            headA = headA->next;
            lenA--;
        }

        // Update the head to match the length of A and B
        while(lenB > lenA)
        {
            headB = headB->next;
            lenB--;
        }

        // finding the intersection point
        while(headA != nullptr || headB != nullptr)
        {
            if(headA == headB)
            {
                return headA;
            }
            headA = headA->next;
            headB = headB->next;
        }
        return nullptr;
    }
};

File: Intersectionof2lists.java
// Time Complexity : O(N) 3 pass
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
public class Intersectionof2lists {

    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {

        if(headA == null || headB == null){
            return null;
        }

        //Finding the lengths of lists

        ListNode l1 = headA;
        ListNode l2 = headB;

        int length1 = 0;
        int length2 = 0;

        while(l1 != null){
            length1++;
            l1 = l1.next;
        }

        while(l2 != null){
            length2++;
            l2 = l2.next;
        }

        //bringing larger list same distance from intersection as the smaller list
        l1 = headA;
        l2 = headB;
        if(length1 > length2){
            while(length1 > length2){
                length1--;
                l1 = l1.next;
            }
        }else{
            while(length2 > length1){
                length2--;
                l2 = l2.next;
            }
        }

        // Now both list equidistant from intersection

        while(l1 != l2){
            l1 = l1.next;
            l2 = l2.next;
        }

        return l1;
    }
}
